File: go/proof.go
package specdecode

import "math"

// InducedFromDists is the closed-form distribution of the token a speculative
// step emits, given the target distribution p and draft distribution q at that
// position. A token t is emitted either because the draft proposed it and it was
// accepted (mass q[t]*accept[t]) or because some proposal was rejected (total
// mass sum(q*(1-accept))) and the residual sample landed on t. Summed, this
// equals p exactly — the exactness theorem the tests assert.
func InducedFromDists(p, q []float64) []float64 {
	n := len(p)
	accepted := make([]float64, n)
	rejectionMass := 0.0
	for i := 0; i < n; i++ {
		a := AcceptProbability(p[i], q[i])
		accepted[i] = q[i] * a
		rejectionMass += q[i] * (1.0 - a)
	}
	r := Residual(p, q)
	out := make([]float64, n)
	for i := 0; i < n; i++ {
		out[i] = accepted[i] + rejectionMass*r[i]
	}
	return out
}

// InducedNextTokenDistribution is the closed-form distribution of the next token
// a speculative step emits from prefix, which the theorem says equals
// target.NextDist(prefix).
func InducedNextTokenDistribution(target, draft Model, prefix []int) []float64 {
	return InducedFromDists(target.NextDist(prefix), draft.NextDist(prefix))
}

// MaxTotalVariation is the largest total-variation distance between the induced
// emission distribution and the target distribution across a set of prefixes. A
// correct implementation returns ~0 (floating-point noise only).
func MaxTotalVariation(target, draft Model, prefixes [][]int) float64 {
	worst := 0.0
	for _, prefix := range prefixes {
		induced := InducedNextTokenDistribution(target, draft, prefix)
		p := target.NextDist(prefix)
		tv := 0.0
		for i := range p {
			tv += math.Abs(induced[i] - p[i])
		}
		tv *= 0.5
		if tv > worst {
			worst = tv
		}
	}
	return worst
}

File: rust/src/proof.rs
//! The correctness proof, made executable: the closed-form distribution the
//! sampler emits, which the theorem says equals the target's.

use crate::models::Model;
use crate::speculative::{accept_probability, residual};

/// Closed-form distribution of the token a speculative step emits, given target
/// distribution `p` and draft distribution `q` at that position. Summed over the
/// accepted and residual paths this equals `p` exactly — the exactness theorem.
pub fn induced_from_dists(p: &[f64], q: &[f64]) -> Vec<f64> {
    let mut accepted: Vec<f64> = Vec::with_capacity(p.len());
    let mut rejection_mass = 0.0;
    for (pi, qi) in p.iter().zip(q) {
        let a = accept_probability(*pi, *qi);
        accepted.push(qi * a);
        rejection_mass += qi * (1.0 - a);
    }
    let r = residual(p, q);
    accepted
        .iter()
        .zip(&r)
        .map(|(a, ri)| a + rejection_mass * ri)
        .collect()
}

/// The closed-form distribution of the next token a speculative step emits from
/// `prefix`, which the theorem says equals `target.next_dist(prefix)`.
pub fn induced_next_token_distribution(
    target: &dyn Model,
    draft: &dyn Model,
    prefix: &[usize],
) -> Vec<f64> {
    induced_from_dists(&target.next_dist(prefix), &draft.next_dist(prefix))
}

/// Largest total-variation distance between the induced emission distribution and
/// the target distribution across a set of prefixes. A correct implementation
/// returns ~0 (floating-point noise only).
pub fn max_total_variation(target: &dyn Model, draft: &dyn Model, prefixes: &[Vec<usize>]) -> f64 {
    let mut worst: f64 = 0.0;
    for prefix in prefixes {
        let induced = induced_next_token_distribution(target, draft, prefix);
        let p = target.next_dist(prefix);
        let tv: f64 = 0.5
            * induced
                .iter()
                .zip(&p)
                .map(|(a, b)| (a - b).abs())
                .sum::<f64>();
        worst = worst.max(tv);
    }
    worst
}
